### PR TITLE
Add a flaky test quarantine job

### PR DESCRIPTION
**Pull request by blindpirate**
This job will be in ReadyForRelease stage, which runs all flaky
tests skipped at most 5 times.

### CODE REVIEW

File: .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    testsWithOs.forEach { testCoverage ->
+        val extraParameters = functionalTestExtraParameters("FlakyTestQuarantine", os, testCoverage.testJvmVersion.major.toString(), testCoverage.vendor.name)
+        val parameters = (
+            buildToolGradleParameters(true) +
+                listOf("-PflakyTestQuarantine") +
+                listOf(extraParameters) +
+                functionalTestParameters(os)
+            ).joinToString(separator = " ")
+        steps {
```
Comment (wolfs) [inline .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt:33]: So we'd run one step for each type of test coverage? How many steps do we end up here?

And I suppose we run those steps without test distribution. right?

File: .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import model.Stage
+
+class FlakyTestQuarantine(model: CIBuildModel, stage: Stage, os: Os) : BaseGradleBuildType(stage = stage, init = {
+    id("${model.projectId}_FlakyQuarantine_${os.name.capitalize()}")
+    name = "Flaky Test Quarantine - ${os.name.capitalize()}"
+    description = "Run all flaky tests skipped multiple times"
+
+    applyDefaultSettings(os, BuildToolBuildJvm, 60)
+
+    val testsWithOs = model.stages.flatMap { it.functionalTests }.filter { it.os == os }
```
Comment (wolfs) [inline .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt:23]: I don't think we should do all stages here, probably only the stages before ready for release, right?

File: build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts [hunk 3]
```diff
@@ -220,7 +220,13 @@ fun Test.configureRerun() {
     }
 }
 
-fun Test.determineMaxRetry() = if (project.name in listOf("smoke-test", "performance", "build-scan-performance")) 1 else 2
+fun Project.isFlakyTestQuarantine() = providers.gradleProperty("flakyTestQuarantine").isPresent
+
+fun Test.determineMaxRetry() = when {
+    project.isFlakyTestQuarantine() -> 4
+    project.name in listOf("smoke-test", "performance", "build-scan-performance") -> 1
+    else -> 2
```
Comment (wolfs) [inline build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts:228]:
```suggestion
    else -> 1
```

Should we reduce the number again to what it was before?

File: .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt (hunk 1 above)
Comment (blindpirate) [reply .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt:33]: I managed to generate TC config with old version of `pom`. There will be 45 steps (including ReadyForRelease) or 33 steps (not including ReadyForRelease)

Comment (blindpirate) [reply .teamcity/src/main/kotlin/configurations/FlakyTestQuarantine.kt:41]: After excluding experimental stages, there're only 9 steps. which seems reasonable. 

Yes, it's without TD, let's see how it works.

File: build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts [hunk 4]
```diff
@@ -230,7 +230,12 @@ fun Test.configureRerun() {
     }
 }
 
-fun Test.determineMaxRetry() = if (project.name in listOf("smoke-test", "performance", "build-scan-performance")) 1 else 2
+fun Project.isFlakyTestQuarantine() = providers.gradleProperty("flakyTestQuarantine").isPresent
```
Comment (wolfs) [inline build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts:233]: This property should go into `BuildParams`.

**Review by wolfs — COMMENTED** [review]
Looks good, I left one comment.

**Comment by blindpirate** [comment]
@bot-gradle test qf

File: build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts (hunk 4 above)
Comment (blindpirate) [reply build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts]: Done.

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Pull request merged**